pkg/dockerregistry/server/util.go

```diff
@@ -77,13 +79,16 @@ func getImportContext(ctx context.Context, ref *imageapi.DockerImageReference, s
 		cred = dockerregistry.NewStaticCredentialStore(&auths[0].AuthConfig)
 	}
 
-	var retriever registryclient.RepositoryRetriever
-	retriever = registryclient.NewContext(
+	impctx := registryclient.NewContext(
 		secureTransport, insecureTransport,
 	).WithRequestModifiers(
 		requesttrace.New(ctx, req),
+	).WithAlternateRepositoryStrategy(
+		strategy.NewSimpleLookupICSPStrategy("", icsp),
```

Comment:
This strategy shouldn't be used by the registry. The registry shouldn't ask api-server for a list for ICSPs on every incoming request, it should use informers that use watches to update their cache.